### PR TITLE
StratifiedRules: Integer literals

### LINKED ISSUE #2705: Stratified rules: Integer literals
Add a new stratified rule for translating integers, which uses the new cache introduced in #2648.

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

closes #2705 

In addition, tweaks the interface of `SubstStratifiedRule`, to avoid implementing an unnecessary do-nothing unit method, and adds a previously missing `BOOLEAN` set test.

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

**Review by konnov — APPROVED** [review]
Nice!

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/RewriterImpl.scala [hunk 1]
```diff
@@ -55,6 +58,8 @@ abstract class RewriterImpl(@unused private val ctx: SolverContext) extends Rewr
     case _                                    => "O@"
   }
 
+  val intValueCache: IntValueCache = new IntValueCache
```
[the remark this answers is not in the document]
Comment (thpani) [reply tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/RewriterImpl.scala:61]: _in the front_, i assume

**Review by thpani — APPROVED** [review]
Code LGTM

I'm concerned though that neither the linked issue nor the current PR description provide context for an uninitiated reader (e.g., a future collaborator, ourselves in 5 years from now, ...) to understand why this PR exists in the first place.
I think these should mention the larger narrative under which these changes are introduced.

**Comment by Kukovec** [comment]
> Code LGTM
> 
> I'm concerned though that neither the linked issue nor the current PR description provide context for an uninitiated reader (e.g., a future collaborator, ourselves in 5 years from now, ...) to understand why this PR exists in the first place. I think these should mention the larger narrative under which these changes are introduced.

The issue is part of https://github.com/informalsystems/apalache/milestone/75, and that milestone links back to the relevant ADR. If this is too obscure, I can start adding more context to individual issues, but then you'd likely have the same description literally copy-pasted for each stratified rule.

**Comment by thpani** [comment]
> The issue is part of https://github.com/informalsystems/apalache/milestone/75, and that milestone links back to the relevant ADR.

If you hadn't told me, I wouldn't have found that.
I think it's unlikely that someone in the future the would have the foresight to trace from this PR, to the issue it's addressing, to that issue's milestone, and then from the milestone description to a PR that merges the relevant ADR.

Why not provide a shortcut, for example by linking to the relevant ADR from the PR and/or issue?

> I can start adding more context to individual issues, but then you'd likely have the same description literally copy-pasted for each stratified rule.

Please do.

**Pull request merged**